src/rust/engine/fs/src/store.rs

```diff
@@ -256,54 +279,201 @@ mod local {
+          directory_database,
+          lease_database,
         }),
       })
     }
-  }
 
-  impl super::ByteStore for ByteStore {
-    fn store_bytes(&self, entry_type: EntryType, bytes: Vec<u8>) -> BoxFuture<Digest, String> {
+    pub fn lease_all<Fs: Iterator<Item = Fingerprint>>(
+      &self,
+      fingerprints: Fs,
+    ) -> Result<(), String> {
+      self
+        .inner
+        .env
+        .begin_rw_txn()
+        .map_err(|err| format!("Error making lmdb transaction: {:?}", err))
+        .and_then(|mut txn| {
+          for fingerprint in fingerprints {
+            self.lease(&fingerprint, &mut txn).map_err(|e| {
+              format!("Error leasing fingerprint {}: {:?}", fingerprint, e)
+            })?;
+          }
+          Ok(txn)
+        })
+        .and_then(|txn| {
+          txn.commit().map_err(
+            |e| format!("Error writing lease: {:?}", e),
+          )
+        })
+    }
+
+    fn lease(&self, fingerprint: &Fingerprint, txn: &mut RwTransaction) -> Result<(), lmdb::Error> {
+      let now_since_epoch = time::SystemTime::now()
+        .duration_since(time::UNIX_EPOCH)
+        .expect("Surely you're not before the unix epoch?");
+      let expires_at = (now_since_epoch + time::Duration::from_secs(60 * 60 * 2)).as_secs();
+      let mut buf = [0; 8];
+      LittleEndian::write_u64(&mut buf, expires_at);
+      // This needs to be a standalone variable because otherwise rustfmt will complain.
+      txn.put(
+        self.inner.lease_database,
+        &fingerprint.as_ref(),
+        &buf,
+        WriteFlags::empty(),
+      )
+    }
+
+    ///
+    /// Attempts to shrink the stored files to be no bigger than target_bytes
+    /// (excluding lmdb overhead).
+    ///
+    /// Returns the size it was shrunk to, which may be larger than target_bytes.
+    ///
+    pub fn shrink(&self, target_bytes: usize) -> Result<usize, String> {
+      let mut used_bytes: usize = 0;
+      let mut fingerprints_by_expired_ago = BinaryHeap::new();
+
+      self
+        .inner
+        .env
+        .begin_rw_txn()
+        .and_then(|mut txn| {
+          {
+            self.aged_fingerprints(
+              &txn,
+              self.inner.file_database,
+              &mut used_bytes,
+              &mut fingerprints_by_expired_ago,
+            );
+            self.aged_fingerprints(
+              &txn,
+              self.inner.directory_database,
+              &mut used_bytes,
+              &mut fingerprints_by_expired_ago,
+            );
+          }
+          while used_bytes > target_bytes {
+            let aged_fingerprint = fingerprints_by_expired_ago.pop().expect(
+              "lmdb corruption detected, sum of size of blobs exceeded stored blobs",
+            );
+            if aged_fingerprint.expired_seconds_ago == 0 {
+              // Ran out of expired blobs - everything remaining is leased and cannot be collected.
+              return Err(MapFull);
+            }
+            txn
```

Comment:
Should this also delete the entry from the lease database?

Reply:
Yes! And now it does! :)